### PR TITLE
feat(composer): add data binding component

### PR DESCRIPTION
**Pull request by sheilaXu**
## Overview
- Add data binding component to store a list of bindings in any node
- Update add/remove overlay component UI to follow the new pattern

Add/remove data binding component

https://user-images.githubusercontent.com/9315598/233741848-e5770edf-a595-4a9d-9cc4-0f0b47154502.mov

Add/remove overlay

https://user-images.githubusercontent.com/9315598/233742071-1f0eeef4-0f02-44a9-9a5d-704e5ce4aba9.mov

### Scene Composer
For `scene-composer` package changes specifically, you can preview the component in the published storybook artifact. To do this, wait for the `Publish Storybook` action to complete below.

- Click on the workflow details
- Select the Summary item on the left
- Download the zip file

To run the storybook build locally, you need a local static web server:

```
npm install -g httpserver
cd <Extracted Zip Directory>
httpserver
```

Then open the website http://localhost:8080 to run the doc site.

## Legal
This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

File: packages/scene-composer/package.json [hunk 1]
```diff
@@ -46,7 +46,7 @@
     "convert-svg": "npx @svgr/cli --out-dir src/assets/auto-gen/icons/ --typescript --index-template tools/index-template.js -- src/assets/icons/",
     "release": "run-s compile copy-assets",
     "copy-assets": "copyfiles -e \"**/*.tsx\" -e \"**/*.ts\" -e \"**/*.snap\" -e \"**/*.js\" -e \"**/*.jsx\" -e \"**/*.json\" \"src/**/*\" dist/",
-    "lint": "eslint . --max-warnings=709",
+    "lint": "eslint . --max-warnings=682",
```
Comment (tracy-french) [inline packages/scene-composer/package.json:49]: Great!

File: packages/scene-composer/src/components/panels/AddComponentMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  Component = 'add-component',
+  Overlay = 'add-component-overlay',
+  DataBinding = 'add-component-data-binding',
+}
+
+type AddComponentMenuItem = ToolbarItemOptions & {
+  uuid: ObjectTypes;
+};
+
+const labelStrings: { [key in ObjectTypes]: MessageDescriptor } = defineMessages({
```
Comment (tracy-french) [inline packages/scene-composer/src/components/panels/AddComponentMenu.tsx:31]: Nit: What are you expecting to gain from this type annotation?

File: packages/scene-composer/src/components/panels/AddComponentMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+interface AddComponentMenuProps {
+  onSelect?: (selectedObject: ObjectTypes) => void;
+}
+
+// Note: ObjectType String is used to record metric. DO NOT change existing ids unless it's necessary.
+enum ObjectTypes {
+  Component = 'add-component',
+  Overlay = 'add-component-overlay',
+  DataBinding = 'add-component-data-binding',
+}
```
Comment (tracy-french) [inline packages/scene-composer/src/components/panels/AddComponentMenu.tsx:25]: Nit: Consider using union of string literals instead of enums. They offer the same effect and work in non-TypeScript projects. The TypeScript team has acknowledged it was a mistake to add functionality not supported by JavaScript and now only focuses on adding types to JavaScript.

File: packages/scene-composer/src/layouts/SceneLayout/components/TabbedPanelContainer.tsx [hunk 4]
```diff
@@ -1,6 +1,8 @@
 import React from 'react';
 import { Tabs } from '@awsui/components-react';
 
+import { TABBED_PANEL_CONTAINER_NAME } from '../../../common/internalConstants';
```
Comment (tracy-french) [inline packages/scene-composer/src/layouts/SceneLayout/components/TabbedPanelContainer.tsx:4]: Nit: Consider using module aliases for shared items to reduce coupling and make imports easier to read.

File: packages/scene-composer/src/components/toolbars/common/styledComponents.ts [hunk 5]
```diff
@@ -80,17 +81,19 @@ export const ToolbarItemIcon = styled.div<ToolbarItemIconProps>`
...
   isOpen?: boolean;
-  orientation?: 'horizontal' | 'vertical';
+  orientation?: ToolbarItemOrientation;
+  position: ToolbarMenuPosition;
 }>`
   display: none;
   position: absolute;
-  left: 100%;
-  top: 0;
+  left: ${({ position }) => (position === 'right' ? '100%' : position === 'bottom-left' ? '0' : undefined)};
```
Comment (tracy-french) [inline packages/scene-composer/src/components/toolbars/common/styledComponents.ts:94]: Nit: This nested ternary is a bit hard to read. Consider refactoring.

File: packages/scene-composer/src/components/toolbars/common/ItemContainer.tsx [hunk 6]
```diff
@@ -59,7 +62,9 @@ export const ItemContainer = React.forwardRef<HTMLDivElement, ToolbarItemContain
     const pointerUp = useCallback(
       (e) => {
         onPointerUp?.(e);
-        onItemClick?.(item);
+        if (!item.isDisabled) {
+          onItemClick?.(item);
+        }
```
Comment (tracy-french) [inline packages/scene-composer/src/components/toolbars/common/ItemContainer.tsx:67]: Nit: Consider adding onItemClick existence to if statement to improve readability.

File: packages/scene-composer/src/components/toolbars/common/ItemContainer.spec.tsx [hunk 7]
```diff
@@ -53,6 +53,13 @@ describe('ItemContainer', () => {
     setFeatureConfig({});
   });
 
+  it('should render selected item properly with custom menuHeight', () => {
+    const { container } = render(
+      <ItemContainer item={{ ...baseItem, isSelected: true }} type='action-select' menuHeight='100px' />,
+    );
+    expect(container).toMatchSnapshot();
+  });
+
```
Comment (tracy-french) [inline packages/scene-composer/src/components/toolbars/common/ItemContainer.spec.tsx:62]: Non-blocker: Consider how to test this without using a snapshot. Please fix this forward.

File: packages/scene-composer/src/components/panels/scene-components/DataBindingComponentEditor.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  const removeComponent = useStore(sceneComposerId)((state) => state.removeComponent);
+  const valueDataBindingProvider = useStore(sceneComposerId)(
+    (state) => state.getEditorConfig().valueDataBindingProvider,
+  );
+
+  const onUpdateCallback = useCallback(
+    (componentPartial: Partial<IDataBindingComponentInternal>, replace?: boolean) => {
+      const componentPartialWithRef = { ref: component.ref, ...componentPartial };
+      if (componentPartialWithRef.valueDataBindings && componentPartialWithRef.valueDataBindings.length > 0) {
+        updateComponentInternal(node.ref, componentPartialWithRef as ISceneComponentInternal, replace);
```
Comment (tracy-french) [inline packages/scene-composer/src/components/panels/scene-components/DataBindingComponentEditor.tsx:30]: Nit: Consider how to avoid type assertion.

File: packages/scene-composer/src/components/panels/scene-components/DataBindingComponentEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  component,
+}: IDataBindingComponentEditorProps) => {
+  const sceneComposerId = useContext(sceneComposerIdContext);
+  const updateComponentInternal = useStore(sceneComposerId)((state) => state.updateComponentInternal);
+  const removeComponent = useStore(sceneComposerId)((state) => state.removeComponent);
+  const valueDataBindingProvider = useStore(sceneComposerId)(
+    (state) => state.getEditorConfig().valueDataBindingProvider,
+  );
+
+  const onUpdateCallback = useCallback(
```
Comment (tracy-french) [inline packages/scene-composer/src/components/panels/scene-components/DataBindingComponentEditor.tsx:26]: Non-blocker: There's a lot going on in this function. Please consider refactoring or adding comments.

File: packages/scene-composer/src/components/panels/SceneNodeInspectorPanel.tsx [hunk 10]
```diff
@@ -97,6 +102,26 @@ export const SceneNodeInspectorPanel: React.FC = () => {
...
+          break;
+        }
+        scrollElement = scrollElement.parentElement;
+      }
+
+      scrollElement?.parentElement?.scroll({
+        top: scrollElement.scrollHeight,
+        behavior: 'smooth',
+      });
+    }, 5);
```
Comment (tracy-french) [inline packages/scene-composer/src/components/panels/SceneNodeInspectorPanel.tsx:122]: Non-blocker: This section could benefit from refactoring or comments to aid understanding. Please improve this forward.

File: packages/scene-composer/src/components/panels/ComponentEditMenu.tsx [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+  const handleRemoveAllDataBinding = useCallback(() => {
+    if (currentComponent.type == KnownComponentType.DataBinding) {
+      removeComponent(nodeRef, currentComponent.ref);
+      return;
+    }
+    throw `Remove all data binding not implemented for current component type ${currentComponent.type}`;
+  }, [nodeRef, currentComponent]);
+
+  return menuItems.length > 1 ? (
+    <div style={{ height: '24px' }}>
```
Comment (tracy-french) [inline packages/scene-composer/src/components/panels/ComponentEditMenu.tsx:142]: Nit: Adding a comment to explain how you got 24px would be helpful.

File: packages/scene-composer/src/components/panels/ComponentEditMenu.tsx [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+          valueDataBindings: [...((currentComponent as IDataBindingComponentInternal).valueDataBindings || []), {}],
+        };
+        updateComponentInternal(nodeRef, newComponentPartial);
+        return;
+      }
+      case KnownComponentType.DataOverlay: {
+        const newComponentPartial = {
+          ...currentComponent,
+          valueDataBindings: [
+            ...((currentComponent as IDataOverlayComponentInternal).valueDataBindings || []),
```
Comment (tracy-french) [inline packages/scene-composer/src/components/panels/ComponentEditMenu.tsx:121]: Nit: Consider how to avoid type assertion.

File: packages/scene-composer/src/components/panels/ComponentEditMenu.tsx [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+            { bindingName: generateUUID() },
+          ],
+        };
+        updateComponentInternal(nodeRef, newComponentPartial);
+        return;
+      }
+      default:
+        throw `Add data binding not implemented for current component type ${currentComponent.type}`;
+    }
+  }, [nodeRef, currentComponent]);
```
Comment (tracy-french) [inline packages/scene-composer/src/components/panels/ComponentEditMenu.tsx:131]: Non-blocker: This function could benefit from refactoring into small functions. It's a bit hard to read. Please consider improving this. Same comment applies to other functions this file.

File: packages/scene-composer/src/components/panels/AddComponentMenu.tsx [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        ...(dataBindingComponentEnabled
+          ? [
+              {
+                uuid: ObjectTypes.DataBinding,
+              },
+            ]
+          : []),
+      ].map(mapToMenuItem),
+    [selectedSceneNodeRef, isOverlayComponent, isTagComponent, dataBindingComponentEnabled],
+  );
```
Comment (tracy-french) [inline packages/scene-composer/src/components/panels/AddComponentMenu.tsx:94]: Non-blocker: This one is a bit hard to follow. Please consider refactoring to improve readability.

**Review by tracy-french — APPROVED** [review]
Great work! Thank you for your contribution @sheilaXu! :)

No blocking feedback. There are a few places your work could benefit from refactoring. I would love to see follow-up work to improve those areas.

**Review by haweston — APPROVED** [review]

File: packages/scene-composer/src/components/panels/AddComponentMenu.tsx (hunk 2 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/AddComponentMenu.tsx:31]: I want to make sure all types in ObjectTypes will get a label string. with this type set, there will be error if new types are not added to this strings map

File: packages/scene-composer/src/common/constants.ts [hunk 15]
```diff
@@ -147,9 +147,12 @@ export const DEFAULT_CAMERA_OPTIONS: Pick<THREE.PerspectiveCamera, 'far' | 'fov'
 };
 export const DEFAULT_CAMERA_TARGET: Vector3 = [0, 0, 0];
 export const DEFAULT_TWEEN_DURATION = 500;
+
+/******************************************************************************
+ * Matterport Constants
+ ******************************************************************************/
```
[the remark this answers is not in the document]
Comment (sheilaXu) [reply packages/scene-composer/src/common/constants.ts:153]: yeah we'd want to have separate files for future constants, that's why I added my new constant in the new file. the size of this file is currently fine, so I'll leave this section in it for now

File: packages/scene-composer/src/common/internalConstants.ts [hunk 16]
```diff
@@ -0,0 +1,7 @@
+export const OPTIONS_PLACEHOLDER_VALUE = 'n/a';
+
+/******************************************************************************
+ * Component ClassName Constants
+ ******************************************************************************/
+// This class name is used by inspector panel to scroll
+export const TABBED_PANEL_CONTAINER_NAME = 'tm-sidePanelTabs';
```
[the remark this answers is not in the document]
Comment (sheilaXu) [reply packages/scene-composer/src/common/internalConstants.ts:7]: sure

File: packages/scene-composer/src/components/panels/AddComponentMenu.tsx (hunk 3 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/AddComponentMenu.tsx:25]: this enum is not supposed to be used by outside of scene composer so should be fine to not work in non-TS project.

File: packages/scene-composer/src/components/panels/AddComponentMenu.tsx (hunk 14 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/AddComponentMenu.tsx:94]: ok

File: packages/scene-composer/src/components/panels/ComponentEditMenu.tsx (hunk 11 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/ComponentEditMenu.tsx:142]: tbh, it's from manual verification to see which value looks better...

File: packages/scene-composer/src/components/panels/SceneNodeInspectorPanel.tsx (hunk 10 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/SceneNodeInspectorPanel.tsx:122]: comments added

File: packages/scene-composer/src/components/panels/scene-components/DataBindingComponentEditor.tsx (hunk 9 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/scene-components/DataBindingComponentEditor.tsx:26]: comments added

File: packages/scene-composer/src/components/panels/scene-components/DataBindingComponentEditor.tsx (hunk 8 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/scene-components/DataBindingComponentEditor.tsx:30]: done

File: packages/scene-composer/src/components/toolbars/common/ItemContainer.tsx (hunk 6 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/toolbars/common/ItemContainer.tsx:67]: done

File: packages/scene-composer/src/components/toolbars/common/ItemContainer.spec.tsx (hunk 7 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/toolbars/common/ItemContainer.spec.tsx:62]: done

File: packages/scene-composer/src/components/toolbars/common/styledComponents.ts (hunk 5 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/toolbars/common/styledComponents.ts:94]: done

File: packages/scene-composer/src/layouts/SceneLayout/components/TabbedPanelContainer.tsx (hunk 4 above)
Comment (sheilaXu) [reply packages/scene-composer/src/layouts/SceneLayout/components/TabbedPanelContainer.tsx:4]: I will try it for my next change so I can have more test when adding the alias

File: packages/scene-composer/src/components/panels/ComponentEditMenu.tsx (hunk 12 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/ComponentEditMenu.tsx:121]: as I've mentioned previously, I don't know if there is a way to avoid this kind of type assertion given our current setup

File: packages/scene-composer/src/components/panels/ComponentEditMenu.tsx (hunk 13 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/ComponentEditMenu.tsx:131]: it seems to be fine currently that each function has a couple of switch cases with a couple of statements for different component types. we can see how to refactor when we need to add more cases

**Review by tracy-french — DISMISSED** [review]

File: packages/scene-composer/src/components/panels/ComponentEditMenu.tsx (hunk 11 above)
Comment (TheEvilDev) [reply packages/scene-composer/src/components/panels/ComponentEditMenu.tsx:142]: In general, we should avoid using "px" as a unit in most scenarios. It's a core tenant of responsive design and accessibility. When you have something like a container that's hard coded to 24px high for example, and the user increases their font-size 800% because they have difficulty seeing, it can mess up the layout, or become illegible.

https://betterprogramming.pub/stop-using-the-pixel-unit-in-css-8b8788a1301f -> This article goes into a lot of the reasons you should avoid using this unit of measure.

Also, prefer having a stylesheet to inline styles.

File: packages/scene-composer/src/components/panels/__snapshots__/ComponentEditor.spec.tsx.snap [hunk 17]
```diff
@@ -29,6 +39,16 @@ exports[`ComponentEditor renders correct component should render the "Camera" ed
 </div>
 `;
 
+exports[`ComponentEditor renders correct component should render the "DataBinding" editor correctly 1`] = `
+<div>
+  <div
+    component="[object Object]"
```
Comment (TheEvilDev) [inline packages/scene-composer/src/components/panels/__snapshots__/ComponentEditor.spec.tsx.snap:45]: Snapshot is not comparing correctly, for example different node or component settings won't change the shapshot, and thus you have a false positive.

File: packages/scene-composer/src/components/panels/__tests__/__snapshots__/CommonPanelComponents.spec.tsx.snap [hunk 18]
```diff
@@ -23,7 +23,22 @@ exports[`render correct panels. DynamicSelect should render Dynamic Select corre
...
+  flex-direction: row;
+  -webkit-box-pack: justify;
+  -webkit-justify-content: space-between;
+  -ms-flex-pack: justify;
+  justify-content: space-between;
+}
+
+.c0 {
   border-bottom: 1px solid colorBorderDividerDefault;
 }
```
Comment (TheEvilDev) [inline packages/scene-composer/src/components/panels/__tests__/__snapshots__/CommonPanelComponents.spec.tsx.snap:43]: This is the output from taking a snapshot of styled components, and tends to be very fragile. It is also annoying to customers if they consume our components, and do tests of their own, so we should generally avoid their use entirely, and stick to classic CSS instead. (We could have a discussion with AppKit about something a bit more modern like TailwindCSS or similar, just as long as we limit the use of "css in js" style of solutions.

File: packages/scene-composer/src/components/toolbars/common/styledComponents.ts [hunk 19]
```diff
@@ -39,19 +39,20 @@ export const Icon = styled(PolarisIcon)<ToolbarItemIconProps>`
...
+  Pick<ToolbarItemOptions, 'isDisabled' | 'isSelected'> & { height?: string }
+>`
   flex: none;
   position: relative;
   display: flex;
   align-items: center;
   justify-content: flex-start;
   min-width: 40px;
-  height: 40px;
+  height: ${({ height }) => height || '40px'};
```
Comment (TheEvilDev) [inline packages/scene-composer/src/components/toolbars/common/styledComponents.ts:51]: nit: Avoid the use of pixels as units.

**Review by TheEvilDev — COMMENTED** [review]
We should try to keep CR's smaller than this, this takes a long time to get through, and I'm sure we've missed some things.

File: packages/scene-composer/src/components/panels/AddComponentMenu.tsx [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+  uuid: ObjectTypes;
+};
+
+const labelStrings: { [key in ObjectTypes]: MessageDescriptor } = defineMessages({
+  [ObjectTypes.Component]: { defaultMessage: 'Add component', description: 'Menu Item label' },
+  [ObjectTypes.Overlay]: { defaultMessage: 'Overlay', description: 'Menu Item label' },
+  [ObjectTypes.DataBinding]: { defaultMessage: 'Data binding', description: 'Menu Item label' },
+});
+
+const textStrings = defineMessages({
```
[the remark this answers is not in the document]
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/AddComponentMenu.tsx:37]: a menu item can be an icon without showing text, so I don't want the same type here

File: packages/scene-composer/src/components/panels/AddComponentMenu.tsx [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+    const component: IDataBindingComponentInternal = {
+      ref: THREE.MathUtils.generateUUID(),
+      type: KnownComponentType.DataBinding,
+      valueDataBindings: [{}],
+    };
+
+    addComponentInternal(selectedSceneNodeRef, component);
+  }, [selectedSceneNodeRef, selectedSceneNode]);
+
+  return addComponentMenuItems.length > 1 ? (
```
[the remark this answers is not in the document]
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/AddComponentMenu.tsx:139]: the first item is the "+" icon and if it's the only item, we don't need to show the whole component

File: packages/scene-composer/src/components/panels/CommonPanelComponents.tsx [hunk 22]
```diff
@@ -98,6 +98,13 @@ export const MatrixLabel = styled.div`
   padding-right: 8px;
 `;
 
+const HeaderContinaer = styled.div`
+  width: 100%;
+  display: flex;
+  flex-direction: row;
+  justify-content: space-between;
+`;
+
```
[the remark this answers is not in the document]
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/CommonPanelComponents.tsx:107]: the main component in this file is constructed with styled component so I'd like to keep it consistent

File: packages/scene-composer/src/components/panels/ComponentEditMenu.tsx (hunk 12 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/ComponentEditMenu.tsx:121]: this is a new menu icon added to inspector panel. we have 4 kinds of valueDataBindings right now. don't expect to seeing additional ones in the future and I may revisit that soon, but at least we will still need to check the component type to see which kind of valueDataBindings the current component is using

File: packages/scene-composer/src/components/panels/ComponentEditMenu.tsx (hunk 13 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/ComponentEditMenu.tsx:131]: there are some menu items shareable by multiple component types, and some items are component type specific. so I'd like to keep this for now and think about how to better refactor when we have more components migrating to use this menu

File: packages/scene-composer/src/components/panels/ComponentEditMenu.tsx (hunk 11 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/ComponentEditMenu.tsx:142]: I tested to increase my browser's font size, the height of the parent component of this menu component doesn't change with font size change, so it's using "px"? which means this component should use "px" as well to make them work together

File: packages/scene-composer/src/components/panels/__snapshots__/ComponentEditor.spec.tsx.snap (hunk 17 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/__snapshots__/ComponentEditor.spec.tsx.snap:45]: will fix

File: packages/scene-composer/src/components/panels/__tests__/__snapshots__/CommonPanelComponents.spec.tsx.snap (hunk 18 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/__tests__/__snapshots__/CommonPanelComponents.spec.tsx.snap:43]: we use https://www.npmjs.com/package/jest-styled-components so there should be a problem

Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/__tests__/__snapshots__/CommonPanelComponents.spec.tsx.snap:43]: comparing with the classic CSS that relies on unique class name to scope style definition, I'd prefer using styled-component because we need to be super careful about the names to make them unique, otherwise it's so easy to get override by or override the other component. I've experienced with this for overlay feature where my panel container class name is the same as Grafana's panel container, then my styles are applied to Grafana panels which broke the whole Grafana dashboard.
so unless we setup something that can help solve the unique class name issue, I'd prefer using styled components

File: packages/scene-composer/src/components/panels/scene-components/DataBindingComponentEditor.tsx [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+    (componentPartial: Partial<IDataBindingComponentInternal>, replace?: boolean) => {
+      const componentPartialWithRef = { ref: component.ref, type: component.type, ...componentPartial };
+      // When the data binding component has valueDataBindings left, update the component, otherwise remove
+      // the whole component instead
+      if (componentPartialWithRef.valueDataBindings && componentPartialWithRef.valueDataBindings.length > 0) {
+        updateComponentInternal(node.ref, componentPartialWithRef, replace);
```
[the remark this answers is not in the document]
Comment (sheilaXu) [reply packages/scene-composer/src/components/panels/scene-components/DataBindingComponentEditor.tsx:32]: `updateComponentDataBinding` will need to check the type of component then to determine the shape of the data binding of that component, so for this function in particular, it won't be much cleaner.

File: packages/scene-composer/src/components/toolbars/common/__snapshots__/ItemContainer.spec.tsx.snap [hunk 24]
```diff
@@ -317,6 +315,47 @@ exports[`ItemContainer should render properly and trigger onClick when clicking
...
+  text-decoration: none;
+  cursor: default;
+  pointer-events: none;
+  background-color: colorBackgroundItemSelected;
+  border-top: 1px solid colorBorderDividerDefault;
+}
+
+.c0:first-of-type {
+  border-top: 0;
+}
```
[the remark this answers is not in the document]
Comment (sheilaXu) [reply packages/scene-composer/src/components/toolbars/common/__snapshots__/ItemContainer.spec.tsx.snap:348]: we use https://www.npmjs.com/package/jest-styled-components to avoid snapshot issue

File: packages/scene-composer/src/components/toolbars/common/styledComponents.ts (hunk 19 above)
Comment (sheilaXu) [reply packages/scene-composer/src/components/toolbars/common/styledComponents.ts:51]: same here. need to make the parent component be responsive first

File: packages/scene-composer/src/components/panels/__tests__/__snapshots__/CommonPanelComponents.spec.tsx.snap (hunk 18 above)
Comment (TheEvilDev) [reply packages/scene-composer/src/components/panels/__tests__/__snapshots__/CommonPanelComponents.spec.tsx.snap:43]: This is easily solved by having a prefix: `tm-` on custom classnames we add. We can see even just from consuming polaris the styled-components model of doing styling is not good for library consumers and we absolutely should move away from it entirely.

I'm open to looking at other options though, scss is just sort of a default fallback. I've been playing with TailwindCSS on a personal project which may be an option.

The main reason I'm pushing on this, is it's making the problem worse by continuing to perpetuate the use of the styled components, when we should be getting rid of it. Besides the extra build step and messed up snapshots, it leads to CSS bloat, as often multiple common styles get duplicated, and the cascading effect of CSS is completely overruled. It's honestly no better than just using inline styles, and we don't need a framework for that.

File: packages/scene-composer/src/components/toolbars/common/__snapshots__/ItemContainer.spec.tsx.snap (hunk 24 above)
[the remark this answers is not in the document]
Comment (TheEvilDev) [reply packages/scene-composer/src/components/toolbars/common/__snapshots__/ItemContainer.spec.tsx.snap:348]: That doesn't fix the snapshot issue, because we see the individual styles right here in the snapshot. This doesn't help customers of our components when writing their own tests. We still create a situation where us changing styles can break previously passing tests downstream that would otherwise pass. It forces customers to effectively adopt styled-components even if they aren't using them.

**Review by TheEvilDev — APPROVED** [review]
